tests/integration/test_upgrade.py

```diff
@@ -0,0 +1,257 @@
+    assert: all operations finished without error.
+    """
+    namespace = ops_test.model_name
+
+    def get_wordpress_podspec_pod():
+        for pod in kube_core_client.list_namespaced_pod(namespace=namespace).items:
+            name = pod.metadata.name
+            if name.startswith("wordpress") and not name.endswith("-0"):
+                return name
+        raise KeyError("can not find wordpress pod")
+
+    wordpress_pod = get_wordpress_podspec_pod()
+
+    def kubernetes_exec(cmd):
+        logger.info(f"exec {cmd} on {wordpress_pod}")
+        resp = kubernetes.stream.stream(
+            kube_core_client.connect_get_namespaced_pod_exec,
+            name=wordpress_pod,
+            namespace=namespace,
+            container="wordpress",
+            command=cmd,
+            stderr=True,
+            stdin=False,
+            stdout=True,
+            tty=False,
+        )
+        logger.info("Response: " + resp)
+
+    kubernetes_exec(
+        [
+            "curl",
+            "-sSL",
+            "https://raw.githubusercontent.com/wp-cli/builds/gh-pages/phar/wp-cli.phar",
+            "-o",
+            "/usr/local/bin/wp",
+        ]
+    )
+    kubernetes_exec(["chmod", "+x", "/usr/local/bin/wp"])
+
+    def wp_cli_exec(wp_cli_cmd):
+        kubernetes_exec(wp_cli_cmd + ["--allow-root", "--path=/var/www/html"])
+
+    wp_cli_exec(["wp", "plugin", "activate", "openstack-objectstorage-k8s"])
+    wp_cli_exec(
+        ["wp", "option", "update", "object_storage", json.dumps(swift_config), f"--format=json"]
+    )
+    unit_ip = unit_ip_list[0]
+    client = WordpressClient(
+        host=unit_ip,
+        username="admin",
+        password=default_admin_password,
+        is_admin=True,
+    )
+    media_id = client.upload_media(filename="wordpress.png", content=test_image)["id"]
+    post = client.create_post(title=POST_TITLE, content=POST_CONTENT, featured_media=media_id)
+    client.create_comment(post_id=post["id"], post_link=post["link"], content=POST_COMMENT)
+    for idx, unit_ip in enumerate(unit_ip_list):
+        await screenshot(f"http://{unit_ip}", screenshot_dir / f"wordpress-before-{idx}.png")
+
+
+@pytest.mark.abort_on_fail
+async def test_build_and_upgrade(
+    ops_test: pytest_operator.plugin.OpsTest,
+    application_name,
+    gen_upgrade_test_charm_config,
+    num_units,
+):
+    """
+    arrange: an old version of the WordPress is deployed.
+    act: build WordPress charm from source and upgrade the WordPress charm.
+    assert: all operations finished without error.
+    """
+
+    charm = await ops_test.build_charm(".")
+    await ops_test.model.remove_application(application_name)
+
+    def wordpress_removed():
+        status = subprocess.check_output(
+            ["juju", "status", "-m", ops_test.model_name, "--format", "json"]
+        )
+        return application_name not in json.loads(status)["applications"]
```

Comment:
ops_test doesn't provide other way to know when and application is effectively removed ?

Reply:
`ops_test` has `ops_test.model.get_status` but it's asynchronous, so it can't be used in `ops_test.model.block_until`.